### DISCUSSION TITLE
new configuration language to obsolete genders systemd usage

### DISCUSSION THREAD

**Discussion by baallan** [ovis-hpc/ldms#722] [Ideas]
Genders has reached its limits with ldms (and gone beyond them).
We're looking at creating a new configuration language for ldmsd.

Some features/desires captured from the 5/20 config files discussion.

 * Deprecate the since-v2 TOSS-packaged genders specification of daemon configuration and all the attendant machinery.
   * This machinery was created to address complexity/redundancy of configuring aggregators (producer lists), stores (explicit naming of schema, lack of support for schema wildcards in storage policies, lack of support of match on instance name in storage policies), and updaters (who would think they should need to define these explicitly?). 
   * As of TOSS4, B. Allan to make the genders-based service into a non-conflicting optional package.
     * rename old ldmsd.service gldmsd.service to avoid conflict with new service.
 * Make a new input language without the verbosity of the current input "object assembly language".
 * Use maestro as a completeness-of-concepts development platform.
 * Make either the input to maestro or an output from maestro directly consumable by ldmsd.
 * Avoid requiring daemons to phone a central server when starting up.
 * Allow a central controller to orchestrate ldmsd's after independent startup, at admin's discretion.
 * Be ansible-friendly.

**Comment by morrone** [comment]
Just brainstorming here, but maybe we could have something like "maestro_ctrl --local", where it reads the ldms config file and only takes action to set up the local components. We might want to give that local compnent a name like "ldms" to not confuse sysadmins. But for discussion I'll just call it "ldmslauncher" for now. The ldmslauncher would be a long-lived daemon that is launched by systemd. It would read from /etc/ldmsd.conf or something similar as the default location of the local configuration file (which would be the maestro yaml format once finalized).  LDMS would fork off the required ldmsd processes.

Perhaps we give ldmsd a new way to receive its configuration instructions...for instance ldmslauncher could create a socket pair for each ldmsd that it is going to luanch, and set one send of the socket pair up as file descriptor 3 in the forked ldmsd. ldmsd gets its configuration instructions right through that socket rather than going to etcd like in the remote configuration system.

**Reply by baallan** [reply]
It would seem the proposed maestro_control --local (can we please stop using random vowel dropping in ldms naming?) has an assumption of one daemon per host or a configuration file tied to the result of hostname.  Nothing particularly wrong with that, but we also need something that accommodates multiple ldmsd instances per host, as this is common for 2nd level aggregators and for scaling purposes.  Perhaps add "maestro-control --instance  $i" looks up configuration by a name rather than assuming a host name. This could then be trivially wrapped by "systemctl start ldmsd@%I".

**Comment by tom95858** [comment]
It doesn't make any of those assumptions. You can have an arbitrary number
of daemons on a single host.

On Mon, May 24, 2021 at 12:06 PM Benjamin Allan ***@***.***>
wrote:

> It would seem the proposed maestro_control --local (can we please stop
> using random vowel dropping in ldms naming?) has an assumption of one
> daemon per host or a configuration file tied to the result of hostname.
> Nothing particularly wrong with that, but we also need something that
> accommodates multiple ldmsd instances per host, as this is common for 2nd
> level aggregators and for scaling purposes. Perhaps add "maestro-control
> --instance $i" looks up configuration by a name rather than assuming a host
> name. This could then be trivially wrapped by "systemctl start ldmsd@%I".
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ovis-hpc/ovis/discussions/722#discussioncomment-777918>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVTPXH72G3FN5ZHTZOVISTTPKIS3ANCNFSM45H36EJA>
> .
>


-- 
Thomas Tucker, President, Open Grid Computing, Inc.

**Reply by baallan** [reply]
yes, but how about an "arbitrary number of named configurations (each of which may include several daemons)" on a single host, for independent management of each group? Without having to proliferate separate .service files, in systemd terms.

**Reply by morrone** [reply]
It isn't clear to me why that would be a good idea, but I don't see anything that would necessarily prevent it either.
